### PR TITLE
Adds Interface and Validation Testing for MAM4xx Online Emissions

### PR DESCRIPTION
**Pull request by mjs271**
An existing EAMxx interface (`eamxx_mam_srf_and_online_emissions_process_interface.cpp`) is modified to include the invocation of the "online" emissions of dust and sea salt. The new additions are for:

1. Reading soil erodibility file for dust emissions
2. Reading the marine organic input file for marine organic aerosols
3. Read sst and dust fluxes from the land and ocean model via surface coupler
4. Invoke dust and sea salt emissions codes from MAM4xx

Since the CMake option, `SCREAM_ENABLE_MAM`, is always `ON`, these codes are always compiled with the EAMxx codebase, but this interface can be turned off/on using the namelist flags. 

### Output variables
The output variable names are "constituent_fluxes" and interstitial aerosols whose emissions are computed:
`nacl_a1`, `mom_a1`, `num_a1`, `nacl_a2`, `mom_a2`, `num_a2`, `dust_a3`, `nacl_a3`, `mom_a3`,`num_a3`, `mom_a4`, `num_a4`

Note that the "constituent_fluxes" indices are exactly the same as EAM's `state%q`. 

### Testing
We have existing tests for this interface, so we didn't add any new tests. We have a suite of validation tests in the MAM4xx repo to validate the online emission processes.

The `srf_online_emiss` process that invokes online emissions codes has to be invoked with the `constituent_fluxes` process to modify interstitial aerosols. Otherwise, only the `constituent_fluxes` array will be modified. These processes are added as an EAMxx process with the ability to turn it on/off using the labels `mam4_srf_online_emiss` and `mam4xx-constituent_fluxes` in the `namelist_scream.xml` file. The `atmchange` command to invoke these processes in a CIME simulation is:

```
./atmchange physics::atm_procs_list="mam4_constituent_fluxes,mac_aero_mic,rrtmgp,mam4_srf_online_emiss"
```

To run a CIME simulation, we have created a test modifier `mam4xx-srf_online_emiss_constituent_fluxes`:

```
cd cime/scripts
./create_test SMS_D_P32x1.ne4pg2_oQU480.F2010-SCREAMv1-MPASSI.<machine>_<compiler>.scream-mam4xx-srf_online_emiss_constituent_fluxes -t <test id> -p <project id>
```

Where,
`<machine>` is the machine name
`<compiler>` is the compiler to use on the machine
`<test id>` is a unique string test identifier
`<project id>` is the allocation project to charge

### New input data
We have added 6 files for each of the three grid resolutions:
`ne2np4`, `ne4pg2`, `ne30pg2`

for soil erodibility (3 files) and marine organic aerosols (3 files)

For finer resolutions, the code will interpolate the `ne30pg2` grid. All the files are on the E3SM input data server.

## Figures from model analysis
Aerosol burdens from surface emissions only simulation at ne30pg2 resolution. We have added prescribed emissions plots to show that they remain unchanged (plots generated by Meng):
image

NOTE: There is an initial unphysical spike in dust burden when we use ne4 resolution. @kaizhangpnl found that that the spike may be due to the huge jump in mean u10 velocity between 1st and 2nd time step (see his comments below). Here are plots from ne4 simulation (generated by Chandru)
image

### Standalone Test Timings (Compy):
We will upload the timings once the PR is reviewed and approved

### CODE REVIEW

**Comment by singhbalwinder** [comment]
Thanks, @mjs271 ! 

Reviewers: I will let you know when it is ready for review. We need to run some tests, which might bring in more changes.

**Comment by singhbalwinder** [comment]
Tagging @susburrows.

File: components/eamxx/cime_config/namelist_defaults_scream.xml
Comment (singhbalwinder) [inline components/eamxx/cime_config/namelist_defaults_scream.xml:316]: This block was added twice (hence repeated), possibly due to 2 PRs merging the same change. We are removing it here.

File: components/eamxx/src/physics/mam/eamxx_mam_microphysics_process_interface.cpp
Comment (singhbalwinder) [inline components/eamxx/src/physics/mam/eamxx_mam_microphysics_process_interface.cpp:54]: These mods are due to the mismatch in the mam4xx submodule. Once microphysics PR is merged, we will rebase to remove these mods.

File: components/eamxx/src/physics/mam/impl/mam4_amicphys.cpp
Comment (singhbalwinder) [inline components/eamxx/src/physics/mam/impl/mam4_amicphys.cpp:680]: These mods are due to the mismatch in the mam4xx submodule. Once microphysics PR is merged, we will rebase to remove these mods.

**Comment by singhbalwinder** [comment]
Removed WIP to get it tested once. I will WIP it again after the testing.

**Comment by singhbalwinder** [comment]
Looks like the model failed to download a data file:
```
Input file /home/projects/e3sm/scream/data/atm/scream/mam4xx/emissions/ne2np4/dst_ne2np4_c20241028.nc needs to be downloaded.
Attempting to download atm/scream/mam4xx/emissions/ne2np4/dst_ne2np4_c20241028.nc from https://web.lcrc.anl.gov/public/e3sm/inputdata/
```
This file is present on the server at:https://web.lcrc.anl.gov/public/e3sm/inputdata/atm/scream/mam4xx/emissions/ne2np4/

I am not sure why the model can't download it.

**Comment by mjs271** [comment]
> Looks like the model failed to download a data file:
> 
> ```
> Input file /home/projects/e3sm/scream/data/atm/scream/mam4xx/emissions/ne2np4/dst_ne2np4_c20241028.nc needs to be downloaded.
> Attempting to download atm/scream/mam4xx/emissions/ne2np4/dst_ne2np4_c20241028.nc from https://web.lcrc.anl.gov/public/e3sm/inputdata/
> ```
> 
> This file is present on the server at:https://web.lcrc.anl.gov/public/e3sm/inputdata/atm/scream/mam4xx/emissions/ne2np4/
> 
> I am not sure why the model can't download it.

@singhbalwinder 

Stranger yet, the file appears to be on weaver already

```
jschm@weaver11:/home/projects/e3sm/scream/data/atm/scream/mam4xx/emissions/ne2np4$ pwd
/home/projects/e3sm/scream/data/atm/scream/mam4xx/emissions/ne2np4
mjschm@weaver11:/home/projects/e3sm/scream/data/atm/scream/mam4xx/emissions/ne2np4$ ls
cmip6_mam4_bc_a4_surf_ne2np4_2010_clim_c20240726.nc
cmip6_mam4_num_a1_surf_ne2np4_2010_clim_c20240726.nc
cmip6_mam4_num_a2_surf_ne2np4_2010_clim_c20240726.nc
cmip6_mam4_num_a4_surf_ne2np4_2010_clim_c20240726.nc
cmip6_mam4_pom_a4_surf_ne2np4_2010_clim_c20240726.nc
cmip6_mam4_so4_a1_surf_ne2np4_2010_clim_c20240726.nc
cmip6_mam4_so4_a2_surf_ne2np4_2010_clim_c20240726.nc
DMSflux.2010.ne2np4_conserv.POPmonthlyClimFromACES4BGC_c20240726.nc
dst_ne2np4_c20241028.nc
```

**Comment by singhbalwinder** [comment]
>This generally seems good.

Thanks, that's good to hear!

>I think the file reading stuff is becoming a bit overkill in the sense of redundant code. Do you have a plan to address it soon or is it on the backburner? To be clear, it is fine to keep it as-is imo but we should document the need to cleaning this up somewhere, maybe punting to Aaron's IO epic :) ... Thoughts are welcome here (or elsewhere). Just noting here because it jumped to me looking at the code

We agree, and we have plans for it, but it is not a high priority at this time. We will likely work on it in the next quarter sometime.

>I think something along these lines came up by Luca last time: in general, comments are good, but sometimes useless comments are pretty frustrating and distracting. There are ton of examples where the comment is easily inferred from the code by an average coder (not even a c++ coder), indicating good quality of code readability. I felt like there were comments where I didn't need them, but there were no comments where we probably should have them (e.g., around device/host issues --- see major point along these lines). Also, block annotations (or whatever they're called, those ornate lines and stuff) are excessive and distracting

I will parse the code again and remove the comments that look unnecessary. I can remove the block annotations (if I understood correctly), but I will keep some of them so that it is visually easy to see where the next routine started

>Like the previous PR, there is a fair bit of warning activity coming from MAM4xx that we don't like to see in the code. It shouldn't be too hard to fix. It's a nuisance to be sure, but I think it will help you ensure better code and avoid delays. I am not 100% sure how many of these are related to this specific PR

We are aware of it. We will address all warnings before merging (like previous PRs). I don't have very reliable access to a GPU machine, so I tend to do all development on the CPU. I was having issues with PM-GPU when I tried. @mjs271 will help us test it on SNL GPU machines and fix all the errors and warnings.


>MAJOR:

>the code doesn't compile on GPUs. The fix isn't too hard. It's to do with stuff like dust_emis_scalefactor being unusable on GPUs. Recall, there are many limitations on what you can do on GPUs. In general, I strongly encourage you to test your code on a GPU platform before submitting the PR.

>Finally, please rebase the PR on top of current master because there were some recent fixes that you'd want to have in here (e.g., you can't conduct tests on pm-gpu without them)

@mjs271 is working on it. I planned to get it working on GPUs and then ask for reviews. In today's dev call, @AaronDonahue advised us to ask for reviews even if it is not completely ready yet to expedite the review process. 
This was a draft PR; we opened it up so that we could run tests and get a first look into where the issues are.

Yes, I will rebase it once microphysics PR merges. @odiazib has resolved the GPU fails on microphysics PR, so it should merge soon.

Thanks for getting to it so quickly!

**Comment by singhbalwinder** [comment]
>To clarify something about the compilation error on GPUs, is stuff like it evading your testing in the MAM4xx repo?

As you mentioned, we do test MAM4xx on GPUs and CPUs. In this PR, I am pointing to a branch of the MAM4xx submodule that has not been tested yet on GPU. We only invoke tests when a PR is issued. I haven't issued a PR for my MAM4xx branch yet, so it has not been tested on GPUs. I will do that soon. We won't merge it until this PR points to the main branch of MAM4xx, which is always tested on GPUs. I hope it makes sense, but let me know if you have any questions.

**Comment by singhbalwinder** [comment]
@mahf708 and @bartgol: We have tested and fixed all GPU-related issues (pm-gpu with sp and Weaver). How do we trigger the AT testing to see if everything works on the testing machines?

**Comment by susburrows** [comment]
@singhbalwinder : the PR looks good from my side, and I added @kaizhangpnl as a reviewer.  When you have analysis plots ready, please add them in the description as usual.

**Comment by singhbalwinder** [comment]
@kaizhangpnl and @susburrows : We have analyzed the output (see PR description), and it is ready to go from our side. Please let us know if you would like us to add anything else. Thanks.

**Comment by kaizhangpnl** [comment]
> @kaizhangpnl and @susburrows : We have analyzed the output (see PR description), and it is ready to go from our side. Please let us know if you would like us to add anything else. Thanks.

Thanks for the information. Are these results from a simulation at ne30pg2?

**Comment by susburrows** [comment]
Thanks for letting us know, @singhbalwinder , and @kaizhangpnl , thanks for promptly following up to review!

**Comment by singhbalwinder** [comment]
>Are these results from a simulation at ne30pg2?

-Yes. I have added this info in the PR description now.

>Are the plotted fields aerosol burdens or emissions?

- These are aerosol burdens. We can add emissions too (i.e., plotting the constituent_flux array).

>Does the test also include the prescribed emissions?

- Yes, we added online emissions in the same interface as prescribed emissions, so we plotted prescribed emissions (SO2 and DMS) to show that they remain unchanged.

>It would be useful to check the emission and burden of dust, sea salt, and marine organic aerosols.

- That's a good idea. We will add mom plots and also the constituent fluxes for these species.

**Review by kaizhangpnl — APPROVED** [review]
I conducted tests on Chrysalis, and the results show reasonable sea salt and marine organic aerosol emissions, with an associated increase in aerosol burden. However, dust emissions display a spike at the start of the model run. A more detailed analysis shows that the 10-meter winds generated within the land model are significantly stronger than those after the initial two time steps. In contrast, no such temporal discrepancy appears in the surface winds from the atmosphere model. This suggests potential issues with the atmosphere-land coupling, which needs further investigation.

Since the changes in this PR do not impact meteorological variables or land surface properties, I think it's OK to integrate this PR first and address the surface wind problem in ELM separately.

**Comment by kaizhangpnl** [comment]
**10m wind speeds from the land model (elm)**
[u10m_elm.pdf](https://github.com/user-attachments/files/17758452/u10m_elm.pdf)

**Surface wind speeds from atm**
[windsurf_atm.pdf](https://github.com/user-attachments/files/17758456/windsurf_atm.pdf)

**Comment by singhbalwinder** [comment]
Thanks, Kai, for the detailed analysis. We found that the initial spike in the dust burden and emission is only observed at ne4 resolution. At the ne30 resolution, we did not see any spike. Here are the plots generated by Meng at ne30 resolution:
image

Here are some plots generated by Chandru at ne4:
image

File: components/eamxx/src/physics/mam/eamxx_mam_srf_and_online_emissions_functions.hpp
Comment (singhbalwinder) [reply components/eamxx/src/physics/mam/eamxx_mam_srf_and_online_emissions_functions.hpp:15]: We are partially zeroing out the `constituent_fluxes` view (only gasses and aerosols). A deep copy will zero out the entire view. Please let me know if that is incorrect.

File: components/eamxx/src/physics/mam/eamxx_mam_srf_and_online_emissions_process_interface.cpp
Comment (singhbalwinder) [reply components/eamxx/src/physics/mam/eamxx_mam_srf_and_online_emissions_process_interface.cpp:423]: Please see above.

File: components/eamxx/src/physics/mam/eamxx_mam_srf_and_online_emissions_functions.hpp
Comment (singhbalwinder) [reply components/eamxx/src/physics/mam/eamxx_mam_srf_and_online_emissions_functions.hpp:69]: That's a good point. We can move it to the interface.

File: components/eamxx/src/physics/mam/eamxx_mam_srf_and_online_emissions_process_interface.cpp
Comment (singhbalwinder) [reply components/eamxx/src/physics/mam/eamxx_mam_srf_and_online_emissions_process_interface.cpp:50]: I am not so clear on this comment. Can you please elaborate?

File: components/eamxx/src/physics/mam/eamxx_mam_srf_and_online_emissions_functions.hpp
Comment (singhbalwinder) [reply components/eamxx/src/physics/mam/eamxx_mam_srf_and_online_emissions_functions.hpp:15]: 0 to gasses_start_ind are all the constituents added by the default EAMxx model (qv, qc, qi, nc, ni, etc.)

File: components/eamxx/src/physics/mam/eamxx_mam_srf_and_online_emissions_process_interface.cpp
Comment (singhbalwinder) [reply components/eamxx/src/physics/mam/eamxx_mam_srf_and_online_emissions_process_interface.cpp:50]: Ok, we followed SHOC here, but we can always revisit it if we want to change it everywhere.

File: components/eamxx/src/physics/mam/eamxx_mam_srf_and_online_emissions_process_interface.cpp
Comment (singhbalwinder) [reply components/eamxx/src/physics/mam/eamxx_mam_srf_and_online_emissions_process_interface.cpp:518]: I will create an issue and revisit it later, as we might need to repeat some analysis after making this change.

File: components/eamxx/src/physics/mam/readfiles/marine_organics_impl.hpp
Comment (singhbalwinder) [reply components/eamxx/src/physics/mam/readfiles/marine_organics_impl.hpp:28]: We will revisit all file reading next year, so I am leaving it as is for now.

**Comment by singhbalwinder** [comment]
@mahf708 : Thanks again for the review. We have addressed all the comments except for the file reading (as we will revisit it later). I will create an issue for the other comment about using the scaling function. @mjs271 have tested on PM-GPUs and these codes are not generating any new warnings.

This should be ready to merge now.

**Review by odiazib — APPROVED** [review]

**Review by kaizhangpnl — APPROVED** [review]

**Pull request merged**
